stdlib/public/core/HashedCollections.swift.gyb

```diff
@@ -505,9 +500,30 @@ public struct Set<Element : Hashable> :
     return _variantStorage.endIndex
   }
 
+  @_versioned
+  @_transparent
+  internal var _guaranteedNative: Bool {
+    return _canBeClass(Element.self) == 0
+  }
+
   // TODO: swift-3-indexing-model - add docs
   public func index(after i: Index) -> Index {
```

Comment:
Could you move this logic into `variantStorage.index(after:)`, and call that method here?  The idea is that `Set` and `Dictionary` should not be directly working with internals of the storage.